src/java.desktop/share/classes/javax/swing/plaf/metal/MetalBorders.java

```diff
@@ -315,72 +326,75 @@ private void paintUnscaledBorder(Component c, Graphics g,
+            return (c instanceof JInternalFrame
+                    && ((JInternalFrame)c).isSelected());
+        }
+
+        @Override
+        protected boolean isResizable(Component c) {
+            return ((c instanceof JInternalFrame
+                    && ((JInternalFrame) c).isResizable()));
+        }
+    }
+
     /**
-     * Border for a Frame.
-     * @since 1.4
+     * The class represents the border of a {@code JInternalFrame}.
      */
     @SuppressWarnings("serial") // Superclass is not serializable across versions
-    static class FrameBorder extends AbstractBorder implements UIResource {
-        private static final int corner = 14;
-
-        public void paintBorder(Component c, Graphics g, int x, int y,
-            int w, int h) {
+    public static class InternalFrameBorder extends AbstractBorder implements UIResource {
 
-            Color background;
-            Color highlight;
-            Color shadow;
+        private final InternalFrameBorderImpl border;
 
-            Window window = SwingUtilities.getWindowAncestor(c);
-            if (window != null && window.isActive()) {
-                background = MetalLookAndFeel.getPrimaryControlDarkShadow();
-                highlight = MetalLookAndFeel.getPrimaryControlShadow();
-                shadow = MetalLookAndFeel.getPrimaryControlInfo();
-            } else {
-                background = MetalLookAndFeel.getControlDarkShadow();
-                highlight = MetalLookAndFeel.getControlShadow();
-                shadow = MetalLookAndFeel.getControlInfo();
-            }
-
-            g.setColor(background);
-            // Draw outermost lines
-            g.drawLine( x+1, y+0, x+w-2, y+0);
-            g.drawLine( x+0, y+1, x+0, y +h-2);
-            g.drawLine( x+w-1, y+1, x+w-1, y+h-2);
-            g.drawLine( x+1, y+h-1, x+w-2, y+h-1);
+        /**
+         * Constructs a {@code InternalFrameBorder}.
+         */
+        public InternalFrameBorder() {
+            border = new InternalFrameBorderImpl();
+        }
 
-            // Draw the bulk of the border
-            for (int i = 1; i < 5; i++) {
-                g.drawRect(x+i,y+i,w-(i*2)-1, h-(i*2)-1);
-            }
+        @Override
+        public void paintBorder(Component c, Graphics g, int x, int y,
+                                int w, int h) {
+            border.paintBorder(c, g, x, y, w, h);
+        }
 
-            if ((window instanceof Frame) && ((Frame) window).isResizable()) {
-                g.setColor(highlight);
-                // Draw the Long highlight lines
-                g.drawLine( corner+1, 3, w-corner, 3);
-                g.drawLine( 3, corner+1, 3, h-corner);
-                g.drawLine( w-2, corner+1, w-2, h-corner);
-                g.drawLine( corner+1, h-2, w-corner, h-2);
+        @Override
+        public Insets getBorderInsets(Component c, Insets newInsets) {
+            newInsets.set(4, 4, 4, 4);
+            return newInsets;
```

Comment:
For consistency with compositing the implementation, it should be:
```suggestion
            return border.getInsets(newInsets);
```